lsp/nls/src/position.rs

```diff
@@ -0,0 +1,162 @@
+use codespan::ByteIndex;
+use nickel_lang_core::{
+    position::TermPos,
+    term::{RichTerm, SharedTerm, Traverse, TraverseControl},
+};
+use range_map::{Range, RangeMap};
+
+// A term that uses pointer equality to implement Eq.
+#[derive(Clone, Debug)]
+struct RichTermPtr(RichTerm);
+
+impl PartialEq for RichTermPtr {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.pos == other.0.pos && SharedTerm::ptr_eq(&self.0.term, &other.0.term)
+    }
+}
+
+impl Eq for RichTermPtr {}
+
+/// A lookup data structure, for looking up the term at a given position.
+///
+/// Overlapping positions are resolved in favor of the smaller one; i.e., lookups return the
+/// most specific term for a given position.
+#[derive(Clone, Debug)]
+pub struct PositionLookup {
+    table: RangeMap<u32, RichTermPtr>,
+}
+
+impl PositionLookup {
+    /// Create a position lookup table for looking up subterms of `rt` based on their positions.
+    pub fn new(rt: &RichTerm) -> Self {
+        let mut all_ranges = Vec::new();
+
+        rt.traverse_ref(&mut |term: &RichTerm| {
+            if let TermPos::Original(pos) = &term.pos {
+                all_ranges.push((pos.start.0, pos.end.0, term.clone()));
+            }
+            TraverseControl::<()>::Continue
+        });
```

Comment:
To be honest, I didn't have plans to merge everything into a single traversal: it seemed clearer to me to have a traversal for positions, a traversal for definitions/usages, and then a typechecking pass. Would it be better to do them all at once?

Reply:
I agree that it's clearer when reading the code, but it's more costly. In general fast compiler tries to have as few phases/passes as possible - some even succeed in doing everything in a single pass. It probably needs to be measured, but if you `RichTerm` is one of tf-ncl' multi MB generated contract, it might have a non negligible cost to traverse it once or five times. But as always, as long as we haven't benchmarked precisely, who knows :man_shrugging: we can start with multiple traversals, benchmark and then improve though.

In general, some traversals might be of different nature and just can't be mashed together (typically, during program transformation, some transformations are bottom-up while others are top-down). But for the LSP all we do in general seems to be: upon visiting some node, store some additional information somewhere (excluding references resolution). This sounds like it can all be done in the same pass.